### DISCUSSION TITLE
Passing ArrayBuffer from C to JS

### DISCUSSION THREAD

**Discussion by linfan68** [Moddable-OpenSource/moddable#1222] [Q&A]
I'm having trouble passing a data buffer from C to JS through a callback.

Here is the JS code:
```javascript
foo({
  // Other values
  onCaptureDone: (a) => {
     trace('============== capture Done CB =============\n')
     trace(`${a.byteLength} ${a[0]} \n`)
   }
 }
);
```

 and C code
``` C
xsSlot _onCaptureDone;
xsmcGet(_onCaptureDone, xsArg(0), xsID_onCaptureDone);
static uint8_t buffer[10] = {0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF, 0x11, 0x22, 0x33, 0x44};
xsmcSetArrayBuffer(xsVar(0), buffer, 10);
xsCallFunction1(xsAccess(_onCaptureDone), xsUndefined, xsVar(0));
```

 and the output
```
============== capture Done CB =============
10 undefined 
```

And I tried passing string, which works fine:

```C
    // static uint8_t buffer[10] = {0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF, 0x11, 0x22, 0x33, 0x44};
    // xsmcSetArrayBuffer(xsVar(0), buffer, 10);
    xsmcSetString(xsVar(0), "Hello");
    xsCallFunction1(xsAccess(_onCaptureDone), xsUndefined, xsVar(0));
```
output
```
============== capture Done CB =============
Hello
```

Any thoughts?

**Comment by phoddie** [comment]
At first glance, the code you shared looks reasonable. Today is a busy, but I will take a look when I can.

One question for you. There must always be a JavaScript stack frame when invoking a callback. If your C callback is triggered outside of an existing XS callback, you need to surround your XS in C calls with `xsBeginHost` / `xsEndHost`. I don't see that in your C code fragment, but perhaps you didn't include it since it is kind of boiler plate. Here's a typical example:

https://github.com/Moddable-OpenSource/moddable/blob/df88825e8abad3979378b7d74562c015a9d49ebb/modules/io/digital/esp32/digitalbank.c#L382-L385

**Reply by phoddie** [reply]
Oh, the problem is in the JavaScript code. ;)

```
trace(`${a.byteLength} ${a[0]} \n`)`
```

The contents of an `ArrayBuffer` cannot be accessed using array notation (e.g. `a[0]`). That gives `undefined`. Running this with any engine -- `(new ArrayBuffer(10))[0]` -- gives `undefined`. From [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer):

> You cannot directly manipulate the contents of an ArrayBuffer; instead, you create one of the [typed array objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray) or a [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) object which represents the buffer in a specific format, and use that to read and write the contents of the buffer

For example, this wraps the `ArrayBuffer` in a `Uint8Array`:

```js
a = new Uint8Array(a);
trace(`${a.byteLength} ${a[0]} \n`)
```
